File: Questao1.java
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.text.DateFormat;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Scanner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class Lista {
    private List<String> Lista;

    Lista(){
        this.Lista = new ArrayList<>();
    }

    public List<String> getLista() {
        return Lista;
    }

    public void setLista(List<String> lista) {
        Lista = lista;
    }

    public void parseStringToList(String texto){
        // Expressão regular para encontrar strings entre aspas simples
        Pattern pattern = Pattern.compile("'(.*?)'");
        Matcher matcher = pattern.matcher(texto);
        
        // Iterar sobre as correspondências encontradas e adicionar à lista
        while (matcher.find()) {
            Lista.add(matcher.group(1)); // Adiciona o texto capturado entre as aspas simples
        }
        
    }
}

class Personagem {
    private String id;
    private String name;
    private Lista alternate_names;
    private String house;
    private String ancestry;
    private String species;
    private String patronus;
    private boolean hogwartsStaff;
    private String hogwatsStudent;
    private String actorName;
    private boolean alive;
    private LocalDate dateOfBirth;
    private int yearOfBith;
    private String eyeColour;
    private String gender;
    private String hairColor;
    private boolean wizard;



    public Personagem(String id, String name, Lista alternate_names, String house, String ancestry, String species,
            String patronus, boolean hogwartsStaff, String hogwatsStudent, String actorName, boolean alive,
            LocalDate dateOfBirth, int yearOfBith, String eyeColour, String gender, String hairColor, boolean wizard) {
        this.id = id;
        this.name = name;
        this.alternate_names = alternate_names;
        this.house = house;
        this.ancestry = ancestry;
        this.species = species;
        this.patronus = patronus;
        this.hogwartsStaff = hogwartsStaff;
        this.hogwatsStudent = hogwatsStudent;
        this.actorName = actorName;
        this.alive = alive;
        this.dateOfBirth = dateOfBirth;
        this.yearOfBith = yearOfBith;
        this.eyeColour = eyeColour;
        this.gender = gender;
        this.hairColor = hairColor;
        this.wizard = wizard;
    }

    public Personagem() {
        this.id = "";
        this.name = "";
        this.alternate_names = [];
        this.house = "";
        this.ancestry = "";
        this.species = "";
        this.patronus = "";
        this.hogwartsStaff = false;
        this.hogwatsStudent = "";
        this.actorName = "";
        this.alive = false;
        this.dateOfBirth = Date.from(null);
        this.yearOfBith = 0;
        this.eyeColour = "";
        this.gender = "";
        this.hairColor = "";
        this.wizard = false;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Lista getAlternate_names() {
        return alternate_names;
    }

    public void setAlternate_names(Lista alternate_names) {
        this.alternate_names = alternate_names;
    }

    public String getHouse() {
        return house;
    }

    public void setHouse(String house) {
        this.house = house;
    }

    public String getAncestry() {
        return ancestry;
    }

    public void setAncestry(String ancestry) {
        this.ancestry = ancestry;
    }

    public String getSpecies() {
        return species;
    }

    public void setSpecies(String species) {
        this.species = species;
    }

    public String getPatronus() {
        return patronus;
    }

    public void setPatronus(String patronus) {
        this.patronus = patronus;
    }

    public boolean isHogwartsStaff() {
        return hogwartsStaff;
    }

    public void setHogwartsStaff(boolean hogwartsStaff) {
        this.hogwartsStaff = hogwartsStaff;
    }

    public String getHogwatsStudent() {
        return hogwatsStudent;
    }

    public void setHogwatsStudent(String hogwatsStudent) {
        this.hogwatsStudent = hogwatsStudent;
    }

    public String getActorName() {
        return actorName;
    }

    public void setActorName(String actorName) {
        this.actorName = actorName;
    }

    public boolean isAlive() {
        return alive;
    }

    public void setAlive(boolean alive) {
        this.alive = alive;
    }

    public LocalDate getDateOfBirth() {
        return dateOfBirth;
    }

    public void setDateOfBirth(LocalDate dateOfBirth) {
        this.dateOfBirth = dateOfBirth;
    }

    public int getYearOfBith() {
        return yearOfBith;
    }

    public void setYearOfBith(int yearOfBith) {
        this.yearOfBith = yearOfBith;
    }

    public String getEyeColour() {
        return eyeColour;
    }

    public void setEyeColour(String eyeColour) {
        this.eyeColour = eyeColour;
    }

    public String getGender() {
        return gender;
    }

    public void setGender(String gender) {
        this.gender = gender;
    }

    public String getHairColor() {
        return hairColor;
    }

    public void setHairColor(String hairColor) {
        this.hairColor = hairColor;
    }

    public boolean isWizard() {
        return wizard;
    }

    public void setWizard(boolean wizard) {
        this.wizard = wizard;
    }

    public void imprimir(){
        System.out.println("[" + id +"##" + name +"##"+ alternate_names + "##" + house + "##" + ancestry + "##" + species + "##" + patronus + "##" + hogwartsStaff + 
        "##" + hogwatsStudent + "##" + actorName + "##" + alive + "##" + dateOfBirth + "##" + yearOfBith + "##" + eyeColour + "##" + gender + "##" + hairColor + "##" + wizard + "]");
    }

    public Personagem clonar(){
        return this;
    }

    public static String[] ler(String dados){
        String[] atributos = dados.split(";");
    
        return atributos;
    }

}

public class Questao1 {

    public static void preencherVetor(Personagem[] personagens){
        String line;
        String[] atributos;
        Lista alternate_names = new Lista();
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd-MM-yyyy");
        try {
            Scanner sc = new Scanner(new File("characters.csv"));
            int i = 0;
            while(sc.hasNextLine()){
                line = sc.nextLine();
                atributos = Personagem.ler(line);
                
                personagens[i] = new Personagem(atributos[0],atributos[1],alternate_names.getLista(), atributos[3], atributos[4], atributos[5], atributos[6], Boolean.parseBoolean(atributos[7]), atributos[8], atributos[9], 
                Boolean.parseBoolean(atributos[10]),LocalDate.parse(atributos[11], formatter), Integer.parseInt(atributos[12]),atributos[13],atributos[14],atributos[15], Boolean.parseBoolean(atributos[16]));
            }

            sc.close();
        } catch (FileNotFoundException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        Personagem[] personagens = new Personagem[405];
        preencherVetor(personagens);
        String id;

        id = sc.nextLine();
        while(id.equals("FIM") != true){

        }

    }

}

File: Questao1.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include <time.h>

typedef struct {
    char Lista[500];
}Lista;

typedef struct {
    char id[200];
    char name[200];
    Lista alternate_names;
    char house[200];
    char ancestry[200];
    char species[200];
    char patronus[200];
    bool hogwartsStaff;
    char hogwartsStudent[200];
    char actorName[200];
    bool alive;
    time_t dateOfBirth;
    int yearOfBith;
    char eyeColour[200];
    char gender[200];
    char hairColor[200];
    bool wizard;
    
}Personagem;

Personagem construtor(char*, char*, char*, char*, char*, char*, char*, bool, char*, char*, bool, time_t,
int, char*, char*, char*, bool);

char* getId(Personagem*);

Personagem setId(char*);

void imprimir(Personagem*);

char* ler(char*);

void preencherVetor(Personagem*);


int main(){

}

Personagem construtor(char id[], char name[], char alternate_names[], char house[], char ancestry[], char species[], char patronus[], bool hogwartsStaff, char hogwartsStudent[], char actorName[], bool alive, time_t dateOfBirth,
int yerOfBirth, char eyeColour[], char gender[], char hairColor[], bool wizard){
    Personagem P;

    strcpy(P.id, id);
    strcpy(P.name, name);
    strcpy(P.alternate_names.Lista, alternate_names);
    strcpy(P.house, house);
    strcpy(P.ancestry, ancestry);
    strcpy(P.patronus, patronus);
    P.hogwartsStaff = hogwartsStaff;
    strcpy(P.hogwartsStudent, hogwartsStudent);
    strcpy(P.actorName, actorName);
    P.alive = alive;
    P.dateOfBirth = dateOfBirth;
    P.yearOfBith = 10;
    strcpy(P.eyeColour, eyeColour);
    strcpy(P.gender, gender);
    strcpy(P.hairColor, hairColor);
    P.wizard = wizard;

    return P;
}

char* ler(char line[]){
    char tokens[18];
    char separador = ';';

    int num_tokens = 0;
    char *token = strtok(line, separador); // Obtém o primeiro token
    while (token != NULL && num_tokens < 18) {
        tokens[num_tokens++] = token; // Armazena o token no array de tokens
        token = strtok(NULL, separador); // Obtém o próximo token
    }

    return tokens;
}

void imprimir(Personagem *P){
    printf("[ %s ## %s ## %s ## %s ## %s ## %s ## %s ## %s]",
     P->id, P->name, P->alternate_names.Lista, P->house, P->ancestry, P->species, P->patronus,
     P->hogwartsStaff);
}

void PreencherVetor(Personagem personagens[]){
    FILE *arquivo_csv;
    char line[400];

    if(arquivo_csv = fopen("characters.csv", "r") != NULL){
        int i = 0;
        while( fgets(line,400,arquivo_csv) != EOF){
            char atributos[] = ler(line);
            personagens[i] = construtor(atributos[0],atributos[1],atributos[2],atributos[3],atributos[4],
            atributos[5], atributos[6], strcmp(atributos[7], "VERDADEIRO") == 0? true: false, atributos[8], atributos[9], atributos[10],
            atributos[12],atributos[13], atributos[14], atributos[15], atributos[16], atributos[17]);
        }
    }
}